File: examples/simple.erl
#!/usr/bin/env escript
%%! -sname hdr_histogram_simple -pa ebin

-mode(compile).

%
% Simple histogram capture example using Erlang
%

loop(_,0) -> ok;
loop(R,X) -> hdr_histogram:record(R,random:uniform(1000000)), loop(R,X-1).

main(_) ->
    {ok,R} = hdr_histogram:open(1000000,3),
    N = 10000000, %2.2 million/sec on my laptop but YMMV

    %% record a random uniform distribution of 1M data points
    S = os:timestamp(),
    loop(R, N),
    E = os:timestamp(),
    X = timer:now_diff(E,S)/1.0e6,
    Y = case X>1 of true -> N/X; false -> N*X end,
    io:format("Runtime: ~psecs ~.5frps~n", [X,Y]),

    %% print percentiles to stdout as CSV
    hdr_histogram:print(R,csv),

    %% print percentiles to stdout as CLASSIC
    hdr_histogram:log(R,classic,"erlang.hgrm"),

    io:format("Min ~p~n", [hdr_histogram:min(R)]),
    io:format("Mean ~.3f~n", [hdr_histogram:mean(R)]),
    io:format("Median ~.3f~n", [hdr_histogram:median(R)]),
    io:format("Max ~p~n", [hdr_histogram:max(R)]),
    io:format("Stddev ~.3f~n", [hdr_histogram:stddev(R)]),
    io:format("99ile ~.3f~n", [hdr_histogram:percentile(R,99.0)]),
    io:format("99.9999ile ~.3f~n", [hdr_histogram:percentile(R,99.9999)]),
    io:format("Memory Size ~p~n", [hdr_histogram:get_memory_size(R)]),
    io:format("Total Count ~p~n", [hdr_histogram:get_total_count(R)]),

    %% we're done, cleanup any held resources
    hdr_histogram:close(R),

    io:format("Done!\n").

File: examples/simple.exs
# Run with elixir -pa ./ebin examples/simple.exs

#
# Simple histogram capture example using Elixir
#
defmodule Simple do
  def main do
    # create a fresh HDR histogram instance
    {:ok, ref} = :hdr_histogram.open(1000000, 3)

    n = 10000000

    # record a random uniform distribution of 1M data points
    started = :os.timestamp
    loop(ref, n)
    ended = :os.timestamp

    duration = :timer.now_diff(ended, started) / 1.0e6
    rate = case duration > 1 do
       true -> n/duration
       false -> n*duration
    end
    :io.format("Runtime: ~psecs ~.5frps~n", [duration, rate])

    # print percentiles to stdout as CSV
    :hdr_histogram.print(ref, :csv)

    # log percentiles to file as CLASSIC
    :hdr_histogram.log(ref, :classic, 'elixir.hgrm')

    # print other values
    IO.puts "Min #{:hdr_histogram.min(ref)}"
    IO.puts "Mean #{:hdr_histogram.mean(ref)}"
    IO.puts "Median #{:hdr_histogram.median(ref)}"
    IO.puts "Max #{:hdr_histogram.max(ref)}"
    IO.puts "Stddev #{:hdr_histogram.stddev(ref)}"
    IO.puts "99ile #{:hdr_histogram.percentile(ref,99.0)}"
    IO.puts "99.9999ile #{:hdr_histogram.percentile(ref,99.9999)}"
    IO.puts "Memory Size #{:hdr_histogram.get_memory_size(ref)}"
    IO.puts "Total Count #{:hdr_histogram.get_total_count(ref)}"

    # we're done, cleanup any held resources
    :hdr_histogram.close(ref)

    IO.puts "Done!"
  end

  def loop(_ref, 0), do: :ok
  def loop(ref, x) do
    :hdr_histogram.record(ref, :random.uniform(1000000))
    loop(ref, x - 1)
  end
end

Simple.main
